### ISSUE TITLE
PG - not retaining reference particle?

### ISSUE THREAD

**Issue by m9ko** [TuringLang/Turing.jl#1598]
Hello,

I've recently updated Turing from v0.13.0 to v0.15.18 and noticed that `PG()` does not work the same way as before, specifically when it comes to retaining reference particles (I'm using Julia 1.4.2 for both). Using an example from Turing Guide: https://turing.ml/dev/docs/using-turing/guide

```julia
@model function gdemo(x, y)
    s ~ InverseGamma(2, 3)
    m ~ Normal(0, sqrt(s))
    x ~ Normal(m, sqrt(s))
    y ~ Normal(m, sqrt(s))
end

# Use only one particle for PG
c2 = sample(gdemo(1.5, 2), PG(1), 1000)
```

I tested it out with only one particle, where PG should retain its original particle throughout the iterations. This is the case for v0.13.0, but not for v0.15.18 (plot of `m` shown below). I'm wondering if this is due to resetting VarInfo before creating the reference particle (line 282 onward in Turing.jl/src/inference/AdvancedSMC.jl). Please let me know if this is intended or I'm mistaken somewhere, as I'm just a Turing/Julia beginner :')

Old Turing ver:

![image](https://user-images.githubusercontent.com/53410259/116788512-fd27b480-aa77-11eb-8cea-ab25a28d2704.png)

New Turing ver:

![image](https://user-images.githubusercontent.com/53410259/116788547-33653400-aa78-11eb-8094-d183fa98c97e.png)

**Comment by devmotion** [comment]
Thanks for opening the issue and providing this clear and simple example! Recently, we started to extract the particle sampling logic to a separate package AdvancedPS.jl such that it can be used even without Turing. Currently, this is not possible (in a nice way) and both packages still not clearly separated. I assume that during the refactoring we (or probably rather I :smile:) broke something accidentally.

In principle, https://github.com/TuringLang/Turing.jl/blob/46c05895d7e51028211a856c78b1cca60838c5d4/src/inference/AdvancedSMC.jl#L294 should ensure that the reference particle is retained but it seems this is not the case anymore. Maybe because the VarInfo object is reset before. I'll have a closer look at the example and check why it fails.

**Comment by devmotion** [comment]
I just checked the implementation in Turing 0.13, and the problem seems to be that the VarInfo object is reset before the reference particle is created. In 0.13 it was done in the correct order: https://github.com/TuringLang/Turing.jl/blob/98869cf5f4713fde8d1473576ec1afb72314efd9/src/inference/AdvancedSMC.jl#L238-L240

**Comment by devmotion** [comment]
I opened a PR: https://github.com/TuringLang/Turing.jl/pull/1599

**Comment by m9ko** [comment]
Thank you so much for checking!

**Comment by devmotion** [comment]
The issue is fixed if you update to Turing 0.15.19 :slightly_smiling_face:

**Comment by m9ko** [comment]
Just ran it, it works great! Thanks again 👍
